designs/drift-hash-versioning.md

```diff
@@ -0,0 +1,180 @@
+    2. Karpenter will need to populate the `NodeClaim` with values that were used in launching the instances. Drift evaluation will need a point of reference on manual checking.
+
+Pros/Cons
+
+- 👍 Cover general cases of updates to the `NodePool` CRD or the hashing mechanism 
+- 👍 Handles cases where users change `NodePool` resources as Karpenter is updating
+- 👎 Cloud-provider would need to propagate on to the NodeClaim to adopt this method for static drift checking 
+- 👎👎 Negatively impact on performance for Karpenter until all NodeClaims match the `nodepool-hash-version`
+- 👎👎 A high maintenance effort required in supporting individual field evaluation 
+
+### Option 2: Update the NodePool API Version and utilize Conversion webhooks to update the `nodepool-hash`    
+
+The Karpenter APIVersion of the CRDs will be changed each time their is a change as such: 
+
+1. A `NodePool` adds default values to existing fields that are included in the hash calculations
+2. A `NodePool` field is added to the hash calculation with an already-set value
+3. A `NodePool` field is removed from the hash calculations
+
+As part of retrieving the `NodePool` from the API Server, a converion webhook would recaluclate the hash and propagte the new hash to the `NodeClaims`.
+
+Pros/Cons 
+- 👍 No Additional permission required on the Karpenter `clusterrole`
+- 👍 Cover general cases of updates to the `NodePool` CRD or the hashing mechanism 
+- 👎 Maintain conversion webhooks 
+- 👎👎 In the case of adding a new field with defaults, the API version would be bumped for non breaking changes  
+
+### Option 3: Add Versioned Hash Annotations
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+  annotations: 
+    karpenter.sh/nodepool-hash: abcdefj
+    karpenter.sh/nodepool-hash-v1: ghijkl
+    karpenter.sh/nodepool-hash-v2: mnopqr
+    karpenter.sh/nodepool-hash-v3: stuvwx
+    ...
+---
+apiVersion: karpenter.sh/v1beta1
+kind: Nodeclaim 
+metadata:
+  name: default
+  annotations: 
+    karpenter.sh/nodepool-hash: abcdefj
+    karpenter.sh/nodepool-hash-v1: ghijkl
+    karpenter.sh/nodepool-hash-v2: mnopqr
+    karpenter.sh/nodepool-hash-v3: stuvwx
+    ...
+
+```
+
+On a Karpenter upgrade, if the hashing mechanism has been updated to contain a new version of the CRDs, a new annotation hash will be added on to the `NodePool`. The new hash will then be back-propagated to the `NodeClaims` that are owned by `NodePool`. During drift evaluation, Karpenter will make sure that all hashes on the `NodeClaims` match the hash present on `NodePool`. If the values do not match, the `NodeClaims` will be drifted. 
+
+Prior to every release, the Karpenter team will validate to see if a new version of the drift hash will be needed based on changes to the CRDs. Each `NodePool` hash annotation will be versioned. We will need to implement proper testing to catch CRD changes that would be breaking and validate that the drift works as expected.
```

Comment:
> Prior to every release, the Karpenter team will validate to see if a new version of the drift hash will be needed based on changes to the CRDs.

Should we need to validate and see if a new version of the drift hash will be needed manually or this is proposed code change?

Reply:
This will be something that is needed to be updated manually. Whenever a user makes a change to the NodePool under these condition: 

1. A `NodePool` adds default values to existing fields that are included in the hash calculations
2. A `NodePool` field is added to the hash calculation with an already-set value
3. A `NodePool` field is removed from the hash calculations

We will need to bump the version annotation on the CRD.